madanalysis/layout/plotflow.py

```diff
@@ -719,157 +719,113 @@ def DrawMATPLOTLIB(
             upper_scale, lower_scale = None, None
             upper_pdf, lower_pdf = None, None
             scale_unc, pdf_unc = None, None
-            if len(pdfset) != 0:
-                if pdfset["weights"].has_scale:
+            if len(weight_set) != 0:
+                # Get the nominal weight
+                if weight_set["weights"].has_scale:
+                    # Configuration
                     dyn_scale = datasets[ind].dynamic_scale_choice
                     n_point_scale = datasets[ind].n_point_scale_variation
-
-                    # Get the nominal weight
-                    central_scale = pdfset["weights"].central_scale
-                    nominal_loc = (
-                        pdfset["weights"]
-                        .get_scale(
-                            dynamic=dyn_scale, muf=central_scale, mur=central_scale
-                        )
-                        .loc
-                    )
+                    scale_vars = weight_set["weights"].get_scale_vars(point=n_point_scale, dynamic=dyn_scale)
+                    logging.getLogger("MA5").debug("Dyn. scale configuration " + str(dyn_scale) + "; " + str(n_point_scale) + "points")
+                    logging.getLogger("MA5").debug("Scale variations = " + str(scale_vars))
+
+                    # Nominal histo
+                    central_scale = weight_set["weights"].central_scale
+                    nominal_loc = weight_set["weights"].get_scale(dynamic=dyn_scale, muf=central_scale, mur=central_scale).loc
+                    logging.getLogger("MA5").debug("Nominal weight location = " + str(nominal_loc))
                     current_histo = np.squeeze(full_histo[:, nominal_loc])
 
-                    scale_vars = pdfset["weights"].get_scale_vars(
-                        point=n_point_scale, dynamic=dyn_scale
-                    )
-
-                    scale_upper_loc = scale_vars[1].loc
-                    scale_lower_loc = scale_vars[0].loc
-                    upper_scale = np.abs(
-                        np.squeeze(full_histo[:, scale_upper_loc]) - current_histo
-                    )
-                    lower_scale = np.abs(
-                        np.squeeze(full_histo[:, scale_lower_loc]) - current_histo
-                    )
+                # Scale variation envelope
+                if weight_set["weights"].has_scale:
+                    upper_histo = np.copy(current_histo)
+                    lower_histo = np.copy(current_histo)
+                    for x in scale_vars:
+                        scale_histo = np.squeeze(full_histo[:, x.loc])
+                        for i, mybin in enumerate(current_histo):
+                            if upper_histo[i] < scale_histo[i]: upper_histo[i] = scale_histo[i]
+                            if lower_histo[i] > scale_histo[i]: lower_histo[i] = scale_histo[i]
+                    upper_scale = np.abs(upper_histo - current_histo)
+                    lower_scale = np.abs(lower_histo - current_histo)
```

Comment:
for loops are slow; i strongly suggest using numpy operations which are based on c/fortran

Reply:
I agree but I am not a numpy expert. Any idea about how to do it?